web-server/src/content/Service/SystemLogs.tsx

```diff
@@ -16,82 +16,171 @@ import { parseLogLine } from '@/utils/logFormatter';
+  const isInitialLoad = useRef(true);
+  const isSearchingRef = useRef(false);
+  const currentMatchRef = useRef(0);
+  const totalMatchesRef = useRef(0);
+  const highlightedElementsRef = useRef<HTMLElement[]>([]);
+  const isNavigatingRef = useRef(false);
 
-  const scrollDown = () => {
-    containerRef.current.scrollIntoView({ behavior: 'smooth' });
-  };
-
+  // Debounce search query updates
   useEffect(() => {
-    const observer = new IntersectionObserver(
-      (entries) => {
-        entries.forEach((entry) => {
-          if (entry.isIntersecting) {
-            showScrollDownButton.false();
-          } else {
-            showScrollDownButton.true();
-          }
-        });
-      },
-      {
-        threshold: 0
-      }
-    );
-
-    const containerElement = containerRef.current;
+    const timer = setTimeout(() => {
+      setDebouncedSearchQuery(searchQuery);
+    }, 300);
+    return () => clearTimeout(timer);
+  }, [searchQuery]);
 
+  const scrollToBottom = useCallback(() => {
     if (containerRef.current) {
-      observer.observe(containerElement);
+      containerRef.current.scrollTo({
+        top: containerRef.current.scrollHeight,
+        behavior: 'smooth'
+      });
+    }
+  }, []);
+
+  const updateHighlight = useCallback((element: HTMLElement | null) => {
+    if (currentHighlightRef.current) {
+      currentHighlightRef.current.style.backgroundColor = 'yellow';
+      currentHighlightRef.current.style.color = 'black';
+    }
+
+    if (element) {
+      element.style.backgroundColor = 'rgb(255, 161, 10)';
+      element.style.color = 'white';
+      currentHighlightRef.current = element;
+    } else {
+      currentHighlightRef.current = null;
     }
+  }, []);
 
-    return () => {
-      if (containerElement) {
-        observer.unobserve(containerElement);
+  const handleSearch = useCallback((query: string) => {
+    setSearchQuery(query);
+    isSearchingRef.current = !!query;
+  
+    if (!query || query.length < 3) {
+      setCurrentMatch(0);
+      currentMatchRef.current = 0;
+      setTotalMatches(0);
+      totalMatchesRef.current = 0;
+      setHighlightedElements([]);
+      highlightedElementsRef.current = [];
+      updateHighlight(null);
+      return;
+    }
+  }, [updateHighlight]);
+  
+  useLayoutEffect(() => {
+    if (!debouncedSearchQuery || debouncedSearchQuery.length < 3) return;
+  
+    const elements = Array.from(
+      containerRef.current?.querySelectorAll('span[style*="background-color: yellow"]') ?? []
+    ) as HTMLElement[];
+  
+    setHighlightedElements(elements);
+    highlightedElementsRef.current = elements;
+    const newTotalMatches = elements.length;
+    setTotalMatches(newTotalMatches);
+    totalMatchesRef.current = newTotalMatches;
+    
+    if (currentMatchRef.current === 0) {
+      setCurrentMatch(newTotalMatches ? 1 : 0);
+      currentMatchRef.current = newTotalMatches ? 1 : 0;
+    }
+  }, [debouncedSearchQuery]);
```

Comment:
Same feedback as my above comments about reducers and state machines.

This code is too easy to add bugs to.

Reply:
Other easily avoidable problems:
- Direct DOM reads and manipulation
- Too many ref changes tracked and updated

Reply:
Also...
💀 (in this context)
`querySelectorAll('span[style*="background-color: yellow"]')`

-100 points to Gryffindor.